### PR TITLE
Fix #3780: [RTL] Image region selection.

### LINKED ISSUE #3780: Image region selection doesn't work in RTL layouts
**Describe the bug**
Image region selection has issues in RTL layouts. Regions don't seem to be selected when tapping around the image.

**To Reproduce**
1. Open the Image Region Selection Exploration under First Test Topic (requires either playing through Prototype Exploration or unlocking it via the developer options menu)
2. Try tapping around the image. Notice that no planets can be selected.

**Expected behavior**
Planets should be selected in this interaction. RTL shouldn't change how image region selection works.

**Screenshots**
None

**Device**
- Device: Nexus 5X

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation

Fix #3780. This PR fixes Image region selection issues in RTL layouts when tapping around the image.

## Screenshot LTR and RTL
Mobile:
Screenshot_1632187231........Screenshot_1632187141

Screenshot_1632187215.......Screenshot_1632187192

## Screenshot for Tests that pass in Expresso after RTL code.

Screenshot (11)

## Screenshot for Tests that fail in Expresso before RTL code.

Screenshot (12)

**Note**: Not tested on Roboelectric added a todo comment in the test.

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by anandwana001** [comment]
This works for me. Nice solution. 🎉

**Review by anandwana001 — APPROVED** [review]
LGTM

**Review by anandwana001 — CHANGES_REQUESTED** [review]
One thing I missed here is, forcing the parent layout to be ltr, when in rtl the conversation format what we needed for rtl is broken.

**Comment by veena14cs** [comment]
> One thing I missed here is, forcing the parent layout to be ltr, when in rtl the conversation format what we needed for rtl is broken.

Parent layout is containing Image only and Images are not been changed to rtl. I didn't get what is getting broken here?

**Comment by anandwana001** [comment]
Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?

RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461

**Comment by veena14cs** [comment]
> Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?
> 
> RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - [#3780 (comment)](https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461)

Margins are correct but if you see clearly there is white background for the image due to which you are seeing margin difference.

**Comment by veena14cs** [comment]
> > Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?
> > RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - [#3780 (comment)](https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461)
> 
> Margins are correct but if you see clearly there is white background for the image due to which you are seeing margin difference.

Yes there is difference between margins. There was padding so it was showing white background. And I think layout direction doesn't work on `ImageView` to fix this I may need to create one more layout that holds imageview

**Comment by veena14cs** [comment]
@anandwana001 I have fixed parent layout please check. I haven't removed padding as I am not sure if I have to keep padding or not. I have updated screenshots and code PTAL.

**Comment by anandwana001** [comment]
The screenshots are now look correct but didn't know why we have to wrap this in constraint layout.

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (BenHenning) [inline app/src/main/res/layout/image_region_selection_interaction_item.xml]: Why is a ConstraintLayout being used? There's just one child--it seems unnecessary.

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (BenHenning) [inline app/src/main/res/layout/image_region_selection_interaction_item.xml]: This & the ltr below seems like the main fix. Why not fix this within the custom view, instead? It seems preferable to have custom views properly support RTL rather than disabling it at the parent.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks for the quick fix @veena14cs.

Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (veena14cs) [reply app/src/main/res/layout/image_region_selection_interaction_item.xml]: As layout direction is affecting margins if we force it to LTR I added a main container to work for RTL and only layout that hold a child will be changed to LTR.

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (veena14cs) [reply app/src/main/res/layout/image_region_selection_interaction_item.xml]: ImageRegionSelectionInteractionView/ImageView is not supporting layout direction. So I fixed it for the parent layout..  I have removed `android:layoutDirection="ltr"` from `ImageRegionSelectionInteractionView`.

**Comment by veena14cs** [comment]
> Thanks for the quick fix @veena14cs.
> 
> Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?

I tried fixing it but co-ordinates are not getting laid-out properly in RTL. The image region co-ordinates are from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks that the co-ordinates are always towards the right end in RTL which is creating problem.

I tried the solution you discussed with @anandwana001 but it didn't work. So I thought this could be the better solution to force the layout to work as LTR, since we aren't changing images from right-to-left for RTL device.

**Comment by veena14cs** [comment]
> > Thanks for the quick fix @veena14cs.
> > Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?
> 
> I tried fixing it but co-ordinates are not getting laid-out properly in RTL. The image region co-ordinates are from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks that the co-ordinates are always towards the right end in RTL which is creating problem.
> 
> I tried the solution you discussed with @anandwana001 but it didn't work. So I thought this could be the better solution to force the layout to work as LTR, since we aren't changing images from right-to-left for RTL device.

This could get fixed from the customView :). I have updated code PTAL.

**Review by rt4914 — CHANGES_REQUESTED** [review]
LGTM, thanks. 

Can we add a test case for this to make sure if someone removes it the test case will fail.

Also, add one line comment above your solution explaining why we are fixing this to LTR always.

**Comment by veena14cs** [comment]
> LGTM, thanks.
> 
> Can we add a test case for this to make sure if someone removes it the test case will fail.
> 
> Also, add one line comment above your solution explaining why we are fixing this to LTR always.

All the tests are ignored in the `ImageRegionSelectionInteractionViewTest`. Is it only for Roboelectric or Expresso too?

**Comment by veena14cs** [comment]
> LGTM, thanks.
> 
> Can we add a test case for this to make sure if someone removes it the test case will fail.
> 
> Also, add one line comment above your solution explaining why we are fixing this to LTR always.

Added test for rtl and added comment explaining the solution.

**Comment by anandwana001** [comment]
@veena14cs 
`Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing`

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: Rather than updating the ignored test, can we have a new test that specifically tests the issue we are trying to solve in this PR. This is we are not modifying the older tests which are already been ignored. 
Or can we confirm that with the new changes in the ignored test are passing the tests or not?

Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: For me on local the test `android://org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_clickRegion2_region2Clicked()` and `android://org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked()` are passing else all are ignored.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Tests are failing on Robolectric and Bazel 

```
org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest > testImageRegionSelectionInteractionView_rtl_clickRegion3_clickRegion2_region2Clicked FAILED
    expected: NamedRegionClickedEvent(regionLabel=Region 2)
    but was : org.oppia.android.app.utility.DefaultRegionClickedEvent@57b692f9
        at 
```

```
org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest > testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked FAILED
    expected: NamedRegionClickedEvent(regionLabel=Region 3)
    but was : org.oppia.android.app.utility.DefaultRegionClickedEvent@2e649950
        at org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked(ImageRegionSelectionInteractionViewTest.kt:161)
```

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: > @veena14cs
> `Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing`

I have added tests screenshots for RTL specific. I haven't tested on Roboelectric I am unable to test on Arctic I need to check how to run tests on Roboelectric. Can you please check other ignored tests that pass for roboelectric or not as there is todo comment added for all the testcases that points to the issue[ #1611](https://github.com/oppia/oppia-android/issues/1611).

**Comment by veena14cs** [comment]
@anandwana001 please check if other tests failures have same error as the new tests. If it has same errors I might have to ignore this for now and update the code. Need to test this once [#1611](https://github.com/oppia/oppia-android/issues/1611) is fixed.

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: Why are we writing ignored tests? Is it because of #1611 ?

**Comment by veena14cs** [comment]
Tests doesn't work on Roboelectric as of now.

**Comment by anandwana001** [comment]
@veena14cs in that case we can make them espresso only.

File: app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt]:
```suggestion
        // ClickableArea coordinates are not laid-out properly in RTL. The image region coordinates are
        // from left-to-right with an upper left origin and touch coordinates from Android start from the right in RTL mode. Thus, to avoid this situation, force layout direction to LTR in all situations.
```

Cleaned up (needs to be re-wrapped at 100 chars per line though).

File: app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt:116]: This is a really nice & elegant solution.

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: Ah this is unfortunate. It seems like Espresso probably is the only choice then.

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (BenHenning) [reply app/src/main/res/layout/image_region_selection_interaction_item.xml]: Looks like this isn't needed anymore & was removed.

File: app/src/main/res/layout/image_region_selection_interaction_item.xml
Comment (BenHenning) [reply app/src/main/res/layout/image_region_selection_interaction_item.xml]: Looks like a solution was found at the view level & this is now resolved.

**Review by BenHenning — COMMENTED** [review]
Thanks @veena14cs! Just had one nit comment, otherwise the PR looks like a really nice solution to the problem.

File: app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt:116]: Thanks Ben!

**Comment by veena14cs** [comment]
> Thanks @veena14cs! Just had one nit comment, otherwise the PR looks like a really nice solution to the problem.

Accepted the nit changes PTAL thanks.

File: app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt]: Not quite done--please also fix the wrapping.

**Comment by BenHenning** [comment]
One follow-up to the documentation update comment thread. PTAL @veena14cs

File: app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt]: Done.

**Comment by veena14cs** [comment]
> One follow-up to the documentation update comment thread. PTAL @veena14cs

Done.

**Review by rt4914 — APPROVED** [review]
LGTM.

For tests I will defer to @anandwana001 Thanks.

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt]: Actually, given this can we use OppiaTestRule and ``RunOn(TestPlatform.ESPRESSO)`` here instead of ignoring? We don't actually need to ignore tests failing for specific platforms anymore.

**Review by BenHenning — COMMENTED** [review]
Thanks @veena14cs! LGTM though I think we can & should re-enable the tests. See my comment.

**Comment by veena14cs** [comment]
> Thanks @veena14cs! LGTM though I think we can & should re-enable the tests. See my comment.

Thanks @BenHenning I have re-enabled tests and updated screenshots in the PR description.
One test was ignored that is to be moved to Roboelectric as per the message.

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:127]: You had to set the oppia test rule, else the RunOn annotation will not work.

```
  @get:Rule
  val oppiaTestRule = OppiaTestRule()
```

Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:127]: `HelpFragmentTest` for reference

**Review by anandwana001 — CHANGES_REQUESTED** [review]

**Review by BenHenning — APPROVED** [review]
LGTM from code perspective other than the issue @anandwana001 mentioned.

@anandwana001 deferring to you to verify CI is fully passing before merging.

Thanks for quick fix @veena14cs.

File: app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:127]: I tested it without this Rule and I I wonder how it worked. Added it now it shouldn't break anywhere. Thanks for reference.

**Comment by veena14cs** [comment]
> LGTM from code perspective other than the issue @anandwana001 mentioned.
> 
> @anandwana001 deferring to you to verify CI is fully passing before merging.
> 
> Thanks for quick fix @veena14cs.

Thanks @BenHenning.

**Comment by anandwana001** [comment]
Assign me once finalize. LGTM for merge.

**Review by anandwana001 — APPROVED** [review]
LGTM

**Comment by veena14cs** [comment]
> Assign me once finalize. LGTM for merge.

Thanks :)

**Pull request merged**
